### PR TITLE
ROX-13916: Using Search and highlighted fields in the v1/alerts/summary/counts API for better scaled performance and to avoid OOM

### PR DESCRIPTION
**Pull request by clickboo**
## Checklist
- [x] Investigated and inspected CI test results
- [x] Unit test and regression tests added
- [x] ~Evaluated and added CHANGELOG entry if required~
- [x] ~Determined and documented upgrade steps~
- [x] ~Documented user facing changes (create PR based on [openshift/openshift-docs]~(https://github.com/openshift/openshift-docs) and merge into [rhacs-docs](https://github.com/openshift/openshift-docs/tree/rhacs-docs))

If any of these don't apply, please comment below.

## Testing Performed

TODO(replace-me)
Automated + manual on cluster. Scale tests TBD.

### CODE REVIEW

File: central/alert/service/service_impl.go [hunk 1]
```diff
@@ -172,8 +172,29 @@ func (s *serviceImpl) GetAlertsCounts(ctx context.Context, request *v1.GetAlerts
 	if request == nil {
 		request = &v1.GetAlertsCountsRequest{}
 	}
+
+	// TODO: @ksanchet - check correctness. This endpoint could possibly be invoked with any query.
+	// For example, alerts within a specific cluster
+	// in that case, will the code below work? The UI only sends in cluster and or namespaces but
+	// an API client can send in any query
```
Comment (md2119) [inline central/alert/service/service_impl.go:179]: If there is a cluster, category, or severity input query then you must transform the respective query from `AddStrings` to `AddStringsHighlighted`. If not, add the wildcard.

File: central/alert/service/service_impl.go [hunk 2]
```diff
@@ -172,8 +172,29 @@ func (s *serviceImpl) GetAlertsCounts(ctx context.Context, request *v1.GetAlerts
...
+			return nil, err
+		}
+	}
+
+	highlightedQ := search.NewQueryBuilder().AddStringsHighlighted(search.Cluster, search.WildcardString).
+		AddStringsHighlighted(search.Category, search.WildcardString).
+		AddStringsHighlighted(search.Severity, search.WildcardString).
+		ProtoQuery()
+
+	alerts, err := s.dataStore.SearchListAlerts(ctx, search.ConjunctionQuery(requestQ, highlightedQ))
```
Comment (md2119) [inline central/alert/service/service_impl.go:197]: use `Search` that returns `[]search.Result` and collect cluster, severity, and category from the `Matches` field.

File: central/alert/service/service_impl.go (hunk 1 above)
Comment (clickboo) [reply central/alert/service/service_impl.go:179]: I had uncommitted changes I just pushed doing exactly that :-) ...check latest

File: central/alert/service/service_impl.go [hunk 3]
```diff
@@ -62,14 +63,14 @@ var (
...
-	// groupByFunctions provides a map of functions that group slices of ListAlet objects by category or by cluser.
-	groupByFunctions = map[v1.GetAlertsCountsRequest_RequestGroup]func(*storage.ListAlert) []string{
-		v1.GetAlertsCountsRequest_UNSET: func(*storage.ListAlert) []string { return []string{""} },
-		v1.GetAlertsCountsRequest_CATEGORY: func(a *storage.ListAlert) (output []string) {
-			output = append(output, a.GetPolicy().GetCategories()...)
+	// groupByFunctions provides a map of functions that group slices of result objects by category or by cluster.
+	groupByFunctions = map[v1.GetAlertsCountsRequest_RequestGroup]func(result search.Result) []string{
+		v1.GetAlertsCountsRequest_UNSET: func(result search.Result) []string { return []string{""} },
+		v1.GetAlertsCountsRequest_CATEGORY: func(a search.Result) (output []string) {
+			output = append(output, a.Matches[search.Category.String()]...)
```
Comment (md2119) [inline central/alert/service/service_impl.go:70]: The key in `Matches` is the field path in the object like `.policy.name` and sth.

You can get the field paths from the list alerts options map as such,
```
field, ok := {alertMappings}.OptionsMap.Get(search.Category.String())
fieldPath := field.GetFieldPath()
```

File: central/alert/service/service_impl.go [hunk 4]
```diff
@@ -172,8 +173,35 @@ func (s *serviceImpl) GetAlertsCounts(ctx context.Context, request *v1.GetAlerts
...
+		if err != nil {
+			return nil, err
+		}
+		search.ApplyFnToAllBaseQueries(q, func(bq *v1.BaseQuery) {
+			matchFieldQuery, ok := bq.GetQuery().(*v1.BaseQuery_MatchFieldQuery)
+			if !ok {
+				return
+			}
+
+			if matchFieldQuery.MatchFieldQuery.GetField() == search.Cluster.String() {
```
Comment (md2119) [inline central/alert/service/service_impl.go:193]: Is the endpoint limited to accepting cluster queries only? You should do the same for Category and Severity.

File: central/alert/service/service_impl.go [hunk 5]
```diff
@@ -510,16 +538,19 @@ func countAlerts(alerts []*storage.ListAlert, groupByFunc func(*storage.ListAler
 	return
 }
 
-func getMapOfAlertCounts(alerts []*storage.ListAlert, groupByFunc func(alert *storage.ListAlert) []string) (groups map[string]map[storage.Severity]int) {
+func getMapOfAlertCounts(alerts []search.Result, groupByFunc func(alert search.Result) []string) (groups map[string]map[storage.Severity]int) {
 	groups = make(map[string]map[storage.Severity]int)
```
Comment (md2119) [inline central/alert/service/service_impl.go:542]: What is the key of the outermost map?

File: central/alert/service/service_impl.go [hunk 6]
```diff
@@ -510,16 +538,19 @@ func countAlerts(alerts []*storage.ListAlert, groupByFunc func(*storage.ListAler
...
 		for _, g := range groupByFunc(a) {
 			if groups[g] == nil {
 				groups[g] = make(map[storage.Severity]int)
 			}
-
-			groups[g][a.GetPolicy().GetSeverity()]++
+			if len(a.Matches[search.Severity.String()]) == 0 {
+				continue
+			}
+			severity, _ := strconv.Atoi(a.Matches[search.Severity.String()][0])
```
Comment (md2119) [inline central/alert/service/service_impl.go:552]: same comment as above regarding the field path

File: central/alert/service/service_impl.go (hunk 4 above)
Comment (clickboo) [reply central/alert/service/service_impl.go:193]: Good point, yes. The UI uses only cluster and namespace but anyone can invoke that API with any query. Done.

File: central/alert/service/service_impl.go (hunk 5 above)
Comment (clickboo) [reply central/alert/service/service_impl.go:542]: the group (aka the category  name or cluster name  or "") - check the expected test results in the related test file.

File: central/alert/service/service_impl.go (hunk 6 above)
Comment (clickboo) [reply central/alert/service/service_impl.go:552]: ??

File: central/alert/service/service_impl.go (hunk 3 above)
Comment (clickboo) [reply central/alert/service/service_impl.go:70]: Got it

**Review by md2119 — CHANGES_REQUESTED** [review]

File: central/alert/service/service_impl.go [hunk 7]
```diff
@@ -510,16 +561,20 @@ func countAlerts(alerts []*storage.ListAlert, groupByFunc func(*storage.ListAler
...
 			if groups[g] == nil {
 				groups[g] = make(map[storage.Severity]int)
 			}
-
-			groups[g][a.GetPolicy().GetSeverity()]++
+			field, _ := mappings.OptionsMap.Get(search.Severity.String())
+			if len(a.Matches[field.GetFieldPath()]) == 0 {
+				continue
+			}
+			severity, _ := strconv.Atoi(a.Matches[field.GetFieldPath()][0])
```
Comment (md2119) [inline central/alert/service/service_impl.go]: discussed the fix on slack

**Review by md2119 — APPROVED** [review]
Minor comments on logging since we already came across unexpected behavior differences.

File: central/alert/service/service_impl.go [hunk 8]
```diff
@@ -510,16 +561,20 @@ func countAlerts(alerts []*storage.ListAlert, groupByFunc func(*storage.ListAler
...
 	groups = make(map[string]map[storage.Severity]int)
 
 	for _, a := range alerts {
 		for _, g := range groupByFunc(a) {
 			if groups[g] == nil {
 				groups[g] = make(map[storage.Severity]int)
 			}
-
-			groups[g][a.GetPolicy().GetSeverity()]++
+			field, _ := mappings.OptionsMap.Get(search.Severity.String())
```
[the remark this answers is not in the document]
Comment (clickboo) [reply central/alert/service/service_impl.go:573]: actually can use MustGet instead. thoughts?

File: central/alert/service/service_impl.go (hunk 7 above)
Comment (clickboo) [reply central/alert/service/service_impl.go]: That root cause though 😓

File: central/alert/service/service_impl.go [hunk 9]
```diff
@@ -172,8 +179,53 @@ func (s *serviceImpl) GetAlertsCounts(ctx context.Context, request *v1.GetAlerts
 	if request == nil {
 		request = &v1.GetAlertsCountsRequest{}
 	}
+
 	request.Request = ensureAllAlertsAreFetched(request.GetRequest())
-	alerts, err := s.dataStore.ListAlerts(ctx, request.GetRequest())
+	var requestQ *v1.Query
+	if request.GetRequest().GetQuery() == "" {
+		requestQ = search.NewQueryBuilder().AddStringsHighlighted(search.Cluster, search.WildcardString).ProtoQuery()
+	} else {
```
Comment (md2119) [inline central/alert/service/service_impl.go:187]: Drop the if block altogether. When there is no query only cluster values are pulled out without severity and category.

**Pull request merged**